binrw_derive/src/parser/types/imports.rs

```diff
@@ -33,32 +37,64 @@ impl Imports {
             Imports::Tuple(ident, _) => Some(quote! {
                 mut #ident
             }),
+            Imports::Named(args) => type_name.map(|type_name| {
+                let args_ty_name = arg_type_name(type_name);
+                let idents = args.iter().map(|x| &x.ident);
+                quote! {
+                    #args_ty_name {
+                        #(#idents),*
+                    }
+                }
+            }),
         }
     }
 
-    pub fn types(&self) -> TokenStream {
+    pub fn args_type(&self, type_name: &Ident) -> (TokenStream, TokenStream) {
+        let empty = quote! {};
         match self {
-            Imports::None => quote! { () },
+            Imports::None => (quote! { () }, empty),
             Imports::List(_, types) => {
                 let types = types.iter();
-                quote! {
-                    (#(#types,)*)
-                }
+                (
+                    quote! {
+                        (#(#types,)*)
+                    },
+                    empty,
+                )
             }
-            Imports::Tuple(_, ty) => ty.to_token_stream(),
+            Imports::Tuple(_, ty) => (ty.to_token_stream(), empty),
+            Imports::Named(args) => generate_named_arg_type(type_name, args),
         }
     }
 }
 
+fn arg_type_name(ty_name: &Ident) -> Ident {
+    format_ident!("{}BinReadArgs", ty_name, span = Span::mixed_site())
+}
+
+fn generate_named_arg_type(ty_name: &Ident, args: &[NamedImport]) -> (TokenStream, TokenStream) {
+    let fields: Vec<BuilderField> = args.iter().map(Into::into).collect();
+
+    let builder_ident = format_ident!("{}BinReadArgBuilder", ty_name, span = Span::mixed_site());
+    let result_name = arg_type_name(ty_name);
+
+    let type_definition = Builder {
+        builder_name: &builder_ident,
+        result_name: &result_name,
+        fields: &fields,
+    }
+    .generate();
```

Comment:
So far as I can tell, this is the only usage of Builder, which means that the only way to create an argument builder is by deriving BinRead. This means that any manual implementation of BinRead that uses named arguments now requires *manually* creating the relevant builder.

Currently we have no implementations that do this inside of binread, which is why this hasn't already been caught, but I plan creating types for that for both Vec and FilePtr (in the form of my proposed Placement type) so I consider this a massive blocker.

I suppose *technically* we could deal with this by having users create another struct that derives BinRead solely so that we can generate the arguments struct, and then steal that struct for struct we actually want to have arguments for. we could probably even pull that off inside of a macro_rules macro. I still don't like it, but, well, it's something i guess.

Reply:
given the workaround mentioned in the above comment, I'm okay with putting said interface off to another PR. I have edited the OP and am resolving.